src/modules/ecdsa_adaptor/main_impl.h

```diff
@@ -0,0 +1,391 @@
+
+    secp256k1_scalar_set_b32(&sk, seckey32, &overflow);
+    ret &= !overflow;
+    ret &= !secp256k1_scalar_is_zero(&sk);
+    secp256k1_scalar_cmov(&sk, &secp256k1_scalar_one, !ret);
+    secp256k1_scalar_set_b32(&msg, msg32, NULL);
+    /* s' = k⁻¹(m + R.x * x) */
+    ret &= secp256k1_ecdsa_adaptor_sig_sign(&sp, &msg, &k, &r, &sk);
+
+    /* return (R, R', s', dleq_proof) */
+    secp256k1_ecdsa_adaptor_sig_serialize(adaptor_sig162, &r, &rp, &sp, &dleq_proof_e, &dleq_proof_s);
+
+    secp256k1_memczero(adaptor_sig162, 162, !ret);
+    secp256k1_scalar_clear(&k);
+    secp256k1_scalar_clear(&sk);
+
+    return ret;
+}
+
+
+/* Modified secp256k1_ecdsa_sig_verify */
+static int secp256k1_ecdsa_adaptor_sig_verify(const secp256k1_context* ctx, secp256k1_gej *pr, secp256k1_scalar *sigr, secp256k1_scalar *sigs, const secp256k1_ge *pubkey, const secp256k1_scalar *message) {
+    secp256k1_scalar sn, u1, u2;
+    secp256k1_gej pubkeyj;
+
+    if (secp256k1_scalar_is_zero(sigr) || secp256k1_scalar_is_zero(sigs)) {
+        return 0;
+    }
+
+    secp256k1_scalar_inverse_var(&sn, sigs);
+    secp256k1_scalar_mul(&u1, &sn, message);
+    secp256k1_scalar_mul(&u2, &sn, sigr);
+    secp256k1_gej_set_ge(&pubkeyj, pubkey);
+    secp256k1_ecmult(&ctx->ecmult_ctx, pr, &pubkeyj, &u2, &u1);
+    if (secp256k1_gej_is_infinity(pr)) {
+        return 0;
+    }
+    return 1;
+}
+
+int secp256k1_ecdsa_adaptor_verify(const secp256k1_context* ctx, const unsigned char *adaptor_sig162, const secp256k1_pubkey *pubkey, const unsigned char *msg32, const secp256k1_pubkey *enckey) {
+    secp256k1_scalar dleq_proof_s, dleq_proof_e;
+    secp256k1_scalar msg;
+    secp256k1_ge pubkey_ge;
+    secp256k1_ge r, rp;
+    secp256k1_scalar sp;
+    secp256k1_scalar sigr;
+    secp256k1_ge enckey_ge;
+    secp256k1_gej derived_rp;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(adaptor_sig162 != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(enckey != NULL);
+
+    if (!secp256k1_ecdsa_adaptor_sig_deserialize(&r, &sigr, &rp, &sp, &dleq_proof_e, &dleq_proof_s, adaptor_sig162)) {
+        return 0;
+    }
+    if (!secp256k1_pubkey_load(ctx, &enckey_ge, enckey)) {
+        return 0;
+    }
+    /* DLEQ_verify((R', Y, R), dleq_proof) */
+    if(!secp256k1_dleq_verify(&ctx->ecmult_ctx, &dleq_proof_s, &dleq_proof_e, &rp, &enckey_ge, &r)) {
+        return 0;
+    }
+    secp256k1_scalar_set_b32(&msg, msg32, NULL);
+    if (!secp256k1_pubkey_load(ctx, &pubkey_ge, pubkey)) {
+        return 0;
+    }
+
+    /* return R' == s'⁻¹(m * G + R.x * X) */
+    if (!secp256k1_ecdsa_adaptor_sig_verify(ctx, &derived_rp, &sigr, &sp, &pubkey_ge, &msg)) {
+        return 0;
+    }
+    secp256k1_gej_neg(&derived_rp, &derived_rp);
+    secp256k1_gej_add_ge_var(&derived_rp, &derived_rp, &rp, NULL);
+    return secp256k1_gej_is_infinity(&derived_rp);
```

Comment:
🔧 It's confusing to me to have this crucial bit just dangling around here. Can we do this in  `ecdsa_adaptor_sig_verify` instead where the rest of the calculation of `derived_rp` is done. Or maybe just inline the above function here. I don't think it's making anything much clearer for me being separate.

Reply:
The idea behind `secp256k1_ecdsa_adaptor_sig_verify` is that it encapsulates the shared code from [secp256k1_ecdsa_sig_verify](https://github.com/ElementsProject/secp256k1-zkp/blob/cfac088e1b2cd3e4a9c06cca3f37ee21bbfd3ab9/src/ecdsa_impl.h#L207), to make it easier to compare to non-adaptor ECDSA verification. This is also the case for `secp256k1_ecdsa_adaptor_sig_sign` which corresponds to [secp256k1_ecdsa_sig_sign](https://github.com/ElementsProject/secp256k1-zkp/blob/cfac088e1b2cd3e4a9c06cca3f37ee21bbfd3ab9/src/ecdsa_impl.h#L277).

I'm not sure how useful that is though, and perhaps it's better to inline and denote the shared code with a comment instead of pulling it out into a function (or just inline it without any comment about the shared code)?

Reply:
> I'm not sure how useful that is though, and perhaps it's better to inline and denote the shared code with a comment instead of pulling it out into a function

Yes I think this would be better :)